collector/lib/Inotify.cpp

```diff
@@ -0,0 +1,132 @@
+#include "Inotify.h"
+
+#include <filesystem>
+
+namespace collector {
+
+std::ostream& operator<<(std::ostream& os, const Watcher& w) {
+  return os << "Watcher:"
+            << "\n  path: " << w.path
+            << "\n  wd: " << w.wd
+            << "\n  tag: " << w.tag;
+}
+
+int Inotify::AddWatcher(const std::filesystem::path& p, uint32_t flags, int tag) {
+  int wd = inotify_add_watch(fd_, p.c_str(), flags);
+  if (wd < 0) {
+    CLOG(ERROR) << "Unable to watch " << p << ": (" << errno << ") " << StrError();
+    return -1;
+  }
+
+  auto it = FindWatcher(p);
+  if (it == watchers_.end()) {
+    watchers_.emplace_back(wd, p, tag);
+  } else {
+    if (it->wd != wd) {
+      inotify_rm_watch(fd_, it->wd);
+      it->wd = wd;
+    }
+    it->tag = tag;
+  }
+
+  return wd;
+}
+
+int Inotify::AddFileWatcher(const std::filesystem::path& p, int tag) {
+  uint32_t flags = IN_MODIFY | IN_MOVE_SELF | IN_DELETE_SELF | IN_DONT_FOLLOW;
+  CLOG(DEBUG) << "Adding file watcher to " << p;
+  return AddWatcher(p, flags, tag);
+}
+
+int Inotify::AddDirectoryWatcher(const std::filesystem::path& p, int tag) {
+  uint32_t flags = IN_CREATE | IN_DELETE | IN_MOVE_SELF | IN_DELETE_SELF | IN_DONT_FOLLOW;
+  CLOG(DEBUG) << "Adding directory watcher to " << p;
+  return AddWatcher(p, flags, tag);
+}
+
+int Inotify::RemoveWatcher(int wd) {
+  return RemoveWatcher(FindWatcher(wd));
+}
+
+int Inotify::RemoveWatcher(const std::filesystem::path& p) {
+  return RemoveWatcher(FindWatcher(p));
+}
+
+int Inotify::RemoveWatcher(WatcherIterator it) {
+  int res = 0;
+  if (it != watchers_.end()) {
+    res = inotify_rm_watch(fd_, it->wd);
+    watchers_.erase(it, it + 1);
+  }
+  return res;
+}
+
+Inotify::WatcherIterator Inotify::FindWatcher(const std::filesystem::path& needle) {
```

Comment:
Why name the variable `needle`.

Reply:
This probably doesn't matter very much since we won't have many watchers, but maybe use a map to avoid iterating.

Reply:
> Why name the variable `needle`.

Because you could think of the watcher store as a haystack, this is something I've seen before and kinda stuck with me. As an example, from `man strstr`:
```
char *strstr(const char *haystack, const char *needle);
```


> This probably doesn't matter very much since we won't have many watchers, but maybe use a map to avoid iterating.

I left a comment w.r.t. this:
https://github.com/stackrox/collector/blob/ed1540635dcb1b8d1bb2af0130764b9ccb2d5af7/collector/lib/Inotify.h#L103-L108
TL;DR we would need 2 maps, one order by path and another by wd, so instead I just used a vector to KISS.